File: app/src/main/java/com/wolcano/musicplayer/music/widgets/ModelView.java
package com.wolcano.musicplayer.music.widgets;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Point;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.util.AttributeSet;
import android.view.View;

import com.wolcano.musicplayer.music.R;
import com.wolcano.musicplayer.music.utils.ImageUtils;


public class ModelView extends View implements ValueAnimator.AnimatorUpdateListener {



    private final float HELPER_PLAY = 0.0f;
    private final float HELPER_PAUSE = -25.0f;
    private final float DICS_INCR = 0.5f;
    private boolean isPlaying = false;
    private Point dicsPnt = new Point();
    private Point modelCoverPnt = new Point();
    private Handler handler = new Handler();
    private Bitmap dicsBitmap;
    private Bitmap helperBitmap;
    private Drawable aboveLine;
    private Point helperCntPnt = new Point();
    private Point modelCoverCntPnt = new Point();
    private float helperRtte = HELPER_PLAY;
    private int cizgiHeight;
    private int modelBWidth;
    private Point helperPnt = new Point();
    private Point discCntPnt = new Point();
    private Drawable modelB;
    private Matrix dicsMatrix = new Matrix();
    private Matrix modelMatrix = new Matrix();
    private Matrix helperMatrix = new Matrix();
    private ValueAnimator playAnim;
    private ValueAnimator pauseAnim;
    private float dicsRtt = 0.0f;
    private Bitmap modelBitmap;
    private static final long GUN_ZAMN = 50L;



    public ModelView(Context context) {
        this(context, null);
    }




    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        if (changed) {
            initModelViewLayout();
        }
    }
    public ModelView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initModelView();
    }
    public ModelView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    private void initModelView() {
        aboveLine = getResources().getDrawable(R.drawable.play_topline);
        modelB = getResources().getDrawable(R.drawable.play_border);
        dicsBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.album_dics);
        modelBitmap = SongCover.get().loadOval(getContext(),null);
        helperBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.grip);
        cizgiHeight = dpDenPx(1);
        modelBWidth = dpDenPx(1);

        playAnim = ValueAnimator.ofFloat(HELPER_PAUSE, HELPER_PLAY);
        playAnim.setDuration(300);
        playAnim.addUpdateListener(this);
        pauseAnim = ValueAnimator.ofFloat(HELPER_PLAY, HELPER_PAUSE);
        pauseAnim.setDuration(300);
        pauseAnim.addUpdateListener(this);
    }


    public void initHelper(boolean isPlaying) {
        helperRtte = isPlaying ? HELPER_PLAY : HELPER_PAUSE;
        invalidate();
    }
    @Override
    protected void onDraw(Canvas canvas) {
        aboveLine.setBounds(0, 0, getWidth(), cizgiHeight);
        aboveLine.draw(canvas);
        modelB.setBounds(dicsPnt.x - modelBWidth, dicsPnt.y - modelBWidth,
                dicsPnt.x + dicsBitmap.getWidth() + modelBWidth, dicsPnt.y + dicsBitmap.getHeight() + modelBWidth);
        modelB.draw(canvas);
        dicsMatrix.setRotate(dicsRtt, discCntPnt.x, discCntPnt.y);
        dicsMatrix.preTranslate(dicsPnt.x, dicsPnt.y);
        canvas.drawBitmap(dicsBitmap, dicsMatrix, null);
        modelMatrix.setRotate(dicsRtt, modelCoverCntPnt.x, modelCoverCntPnt.y);
        modelMatrix.preTranslate(modelCoverPnt.x, modelCoverPnt.y);
        canvas.drawBitmap(modelBitmap, modelMatrix, null);
        helperMatrix.setRotate(helperRtte, helperCntPnt.x, helperCntPnt.y);
        helperMatrix.preTranslate(helperPnt.x, helperPnt.y);
        canvas.drawBitmap(helperBitmap, helperMatrix, null);
    }

    private void initModelViewLayout() {
        if (getWidth() == 0 || getHeight() == 0) {
            return;
        }

        int unit = Math.min(getWidth(), getHeight()) / 8;
        SongCover.get().setOvalLength(unit * 4);
        dicsBitmap = ImageUtils.chgImage(dicsBitmap, unit * 6, unit * 6);
        modelBitmap = ImageUtils.chgImage(modelBitmap, unit * 4, unit * 4);
        helperBitmap = ImageUtils.chgImage(helperBitmap, unit * 2, unit * 3);

        int discOffsetY = helperBitmap.getHeight() / 2;
        dicsPnt.x = (getWidth() - dicsBitmap.getWidth()) / 2;
        dicsPnt.y = discOffsetY;
        modelCoverPnt.x = (getWidth() - modelBitmap.getWidth()) / 2;
        modelCoverPnt.y = discOffsetY + (dicsBitmap.getHeight() - modelBitmap.getHeight()) / 2;
        helperPnt.x = getWidth() / 2 - helperBitmap.getWidth() / 6;
        helperPnt.y = -helperBitmap.getWidth() / 6;
        discCntPnt.x = getWidth() / 2;
        discCntPnt.y = dicsBitmap.getHeight() / 2 + discOffsetY;
        modelCoverCntPnt.x = discCntPnt.x;
        modelCoverCntPnt.y = discCntPnt.y;
        helperCntPnt.x = discCntPnt.x;
        helperCntPnt.y = 0;
    }

    public void setRoundBitmap(Bitmap bitmap) {
        modelBitmap = bitmap;
        dicsRtt = 0.0f;
        invalidate();
    }
    public void pause() {
        if (!isPlaying) {
            return;
        }
        isPlaying = false;
        handler.removeCallbacks(runnableRtt);
        pauseAnim.start();
    }
    public void start() {
        if (isPlaying) {
            return;
        }
        isPlaying = true;
        handler.post(runnableRtt);
        playAnim.start();
    }



    @Override
    public void onAnimationUpdate(ValueAnimator animation) {
        helperRtte = (float) animation.getAnimatedValue();
        invalidate();
    }
    private int dpDenPx(float dpValue) {
        float density = getContext().getResources().getDisplayMetrics().density;
        return (int) (dpValue * density + 0.5f);
    }
    private Runnable runnableRtt = new Runnable() {
        @Override
        public void run() {
            if (isPlaying) {
                dicsRtt += DICS_INCR;
                if (dicsRtt >= 360) {
                    dicsRtt = 0;
                }
                invalidate();
            }
            handler.postDelayed(this, GUN_ZAMN);
        }
    };


}

File: app/src/main/java/com/wolcano/musicplayer/music/widgets/ModelView.kt
package com.wolcano.musicplayer.music.widgets

import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import androidx.core.content.ContextCompat
import com.wolcano.musicplayer.music.R
import com.wolcano.musicplayer.music.utils.ImageUtils
import com.wolcano.musicplayer.music.widgets.SongCover.loadOval
import com.wolcano.musicplayer.music.widgets.SongCover.setOvalLength
import kotlin.math.min

class ModelView : View, AnimatorUpdateListener {

    private val HELPER_PLAY = 0.0f
    private val HELPER_PAUSE = -25.0f
    private val DISC_INCR = 0.5f
    private var isPlaying = false
    private val discPoint = Point()
    private val modelCoverPnt = Point()
    private var aboveLine: Drawable? = null
    private val helperCntPnt = Point()
    private val modelCoverCntPnt = Point()
    private var helperRtte = HELPER_PLAY
    private var lineHeight = 0
    private var modelBWidth = 0
    private val helperPnt = Point()
    private val discCntPnt = Point()
    private var modelB: Drawable? = null
    private val discMatrix = Matrix()
    private val modelMatrix = Matrix()
    private val helperMatrix = Matrix()
    private var playAnim: ValueAnimator? = null
    private var pauseAnim: ValueAnimator? = null
    private var discRtt = 0.0f
    private var discBitmap: Bitmap? = null
    private var modelBitmap: Bitmap? = null
    private var helperBitmap: Bitmap? = null
    private val delay = 50L

    constructor(context: Context) : super(context, null)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        initModelView()
    }

    constructor(context: Context, attrs: AttributeSet) : this(context, attrs, 0)

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        if (changed) {
            initModelViewLayout()
        }
    }

    private fun initModelView() {

        aboveLine = ContextCompat.getDrawable(context, R.drawable.play_topline)
        modelB = ContextCompat.getDrawable(context, R.drawable.play_border)
        discBitmap = BitmapFactory.decodeResource(resources, R.drawable.album_dics)
        modelBitmap = loadOval(context, null)
        helperBitmap = BitmapFactory.decodeResource(resources, R.drawable.grip)
        lineHeight = dpDenPx(1f)
        modelBWidth = dpDenPx(1f)
        playAnim = ValueAnimator.ofFloat(HELPER_PAUSE, HELPER_PLAY)
        playAnim?.duration = 300
        playAnim?.addUpdateListener(this)
        pauseAnim = ValueAnimator.ofFloat(HELPER_PLAY, HELPER_PAUSE)
        pauseAnim?.duration = 300
        pauseAnim?.addUpdateListener(this)
    }

    fun initHelper(isPlaying: Boolean) {
        helperRtte = if (isPlaying) HELPER_PLAY else HELPER_PAUSE
        invalidate()
    }

    override fun onDraw(canvas: Canvas) {
        aboveLine?.setBounds(0, 0, width, lineHeight)
        aboveLine?.draw(canvas)
        modelB?.setBounds(
            discPoint.x - modelBWidth,
            discPoint.y - modelBWidth,
            discPoint.x + discBitmap!!.width + modelBWidth,
            discPoint.y + discBitmap!!.height + modelBWidth
        )
        modelB?.draw(canvas)
        discMatrix.setRotate(discRtt, discCntPnt.x.toFloat(), discCntPnt.y.toFloat())
        discMatrix.preTranslate(discPoint.x.toFloat(), discPoint.y.toFloat())
        if (discBitmap != null) {
            canvas.drawBitmap(discBitmap, discMatrix, null)
        }
        modelMatrix.setRotate(discRtt, modelCoverCntPnt.x.toFloat(), modelCoverCntPnt.y.toFloat())
        modelMatrix.preTranslate(modelCoverPnt.x.toFloat(), modelCoverPnt.y.toFloat())
        if (modelBitmap != null) {
            canvas.drawBitmap(modelBitmap!!, modelMatrix, null)
        }
        helperMatrix.setRotate(helperRtte, helperCntPnt.x.toFloat(), helperCntPnt.y.toFloat())
        helperMatrix.preTranslate(helperPnt.x.toFloat(), helperPnt.y.toFloat())
        if (helperBitmap != null) {
            canvas.drawBitmap(helperBitmap, helperMatrix, null)
        }
    }


    private fun initModelViewLayout() {
        if (width == 0 || height == 0) {
            return
        }
        val unit = min(width, height) / 8
        setOvalLength(unit * 4)
        discBitmap = ImageUtils.updateImage(discBitmap, unit * 6, unit * 6)
        modelBitmap = ImageUtils.updateImage(modelBitmap, unit * 4, unit * 4)
        helperBitmap = ImageUtils.updateImage(helperBitmap, unit * 2, unit * 3)
        val discOffsetY = helperBitmap!!.height / 2
        discPoint.x = (width - discBitmap!!.width) / 2
        discPoint.y = discOffsetY
        modelCoverPnt.x = (width - modelBitmap?.width!!) / 2
        modelCoverPnt.y = discOffsetY + (discBitmap!!.height - modelBitmap?.height!!) / 2
        helperPnt.x = width / 2 - helperBitmap?.width!! / 6
        helperPnt.y = -helperBitmap?.width!! / 6
        discCntPnt.x = width / 2
        discCntPnt.y = discBitmap!!.height / 2 + discOffsetY
        modelCoverCntPnt.x = discCntPnt.x
        modelCoverCntPnt.y = discCntPnt.y
        helperCntPnt.x = discCntPnt.x
        helperCntPnt.y = 0
    }

    fun setRoundBitmap(bitmap: Bitmap?) {
        modelBitmap = bitmap
        discRtt = 0.0f
        invalidate()
    }

    fun pause() {
        if (!isPlaying) {
            return
        }
        isPlaying = false
        handler.removeCallbacks(runnableRtt)
        pauseAnim?.start()
    }

    fun start() {
        if (isPlaying) {
            return
        }
        isPlaying = true
        handler.post(runnableRtt)
        playAnim?.start()
    }

    override fun onAnimationUpdate(animation: ValueAnimator) {
        helperRtte = animation.animatedValue as Float
        invalidate()
    }

    private fun dpDenPx(dpValue: Float): Int {
        val density = context.resources.displayMetrics.density
        return (dpValue * density + 0.5f).toInt()
    }

    private val runnableRtt: Runnable = object : Runnable {
        override fun run() {
            if (isPlaying) {
                discRtt += DISC_INCR
                if (discRtt >= 360) {
                    discRtt = 0f
                }
                invalidate()
            }
            handler?.postDelayed(this, delay)
        }
    }

}